### PR TITLE
[PCFG-17] CRUD Operations for Creditor Institutions

### PR DESCRIPTION
**Pull request by jacopocarlini**
Implemented POST `/creditorinstitutions`
Implemented PUT `/creditorinstitutions/:id`
Implemented DELETE `/creditorinstitutions/:id`

#### List of Changes

- updated `CreditorInstitutionsController.java` 
   -  added new endpoints
- updated `CreditorInstitutionsService.java` 
  -  implemented business logic
- updated Documentation
  - swagger.json
  - openapi.json
  - README.md
- environment configurations
  - added H2 init scripts
  - modified deploy pipeline

#### Motivation and Context

To introduce new features

#### How Has This Been Tested?

Added Junit Tests

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: api-test/ApiConfig.postman_collection.json [hunk 1]
```diff
@@ -48,6 +48,46 @@
...
+								"type": "text/javascript"
+							}
+						}
+					],
+					"request": {
+						"method": "POST",
+						"header": [],
+						"body": {
+							"mode": "raw",
+							"raw": "{\n  \"creditor_institution_code\": \"0010101010\",\n  \"enabled\": true,\n  \"business_name\": \"Comune di Lorem Ipsum\",\n  \"address\": {\n    \"location\": \"Via delle vie 3\",\n    \"city\": \"Rome\",\n    \"zip_code\": \"00187\",\n    \"country_code\": \"RM\",\n    \"tax_domicile\": \"00111\"\n  },\n  \"fk_int_quadrature\": null,\n  \"psp_payment\": true,\n  \"reporting_ftp\": false,\n  \"reporting_zip\": true\n}",
```
Comment (pasqualespica) [inline api-test/ApiConfig.postman_collection.json:71]: Could be better to use a generic random creditor code , to avoid that two consecutive executions of the collections give differential results ( ex. `"status": 409` ) 

In order to do that we can sue postman `Pre-request `
```
function makeid(length) {
    var result           = '';
    var characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for ( var i = 0; i < length; i++ ) {
      result += characters.charAt(Math.floor(Math.random() * 
 charactersLength));
   }
   return result;
}

pm.globals.set("creditor_institution_code", makeid(35));
```
As next step we shuold extend collection with edge case too, example `creditor_institution_code=""`

File: api-test/ApiConfig.postman_collection.json [hunk 2]
```diff
@@ -91,6 +132,53 @@
...
+								"type": "text/javascript"
+							}
+						}
+					],
+					"request": {
+						"method": "PUT",
+						"header": [],
+						"body": {
+							"mode": "raw",
+							"raw": "{\n    \"creditor_institution_code\": \"0010101010\",\n    \"enabled\": true,\n    \"business_name\": \"Comune di Lorem Ipsum\",\n    \"address\": {\n        \"location\": \"Via delle vie 4\",\n        \"city\": \"Rome\",\n        \"zip_code\": \"00187\",\n        \"country_code\": \"RM\",\n        \"tax_domicile\": \"00111\"\n    },\n    \"fk_int_quadrature\": null,\n    \"psp_payment\": true,\n    \"reporting_ftp\": false,\n    \"reporting_zip\": true\n}",
```
Comment (pasqualespica) [inline api-test/ApiConfig.postman_collection.json:155]: as above ex. we could test 
```
{
    "title": "Creditor Institution not found",
    "status": 404,
    "detail": "No creditor institution found with the provided code"
}
```
too

File: api-test/ApiConfig.postman_collection.json [hunk 3]
```diff
@@ -295,6 +383,53 @@
...
+								"type": "text/javascript"
+							}
+						}
+					],
+					"request": {
+						"method": "DELETE",
+						"header": [],
+						"body": {
+							"mode": "raw",
+							"raw": "{\n  \"creditor_institution_code\": \"1234567890100\",\n  \"enabled\": true,\n  \"business_name\": \"Comune di Lorem Ipsum\",\n  \"address\": {\n    \"location\": \"Via delle vie 3\",\n    \"city\": \"Rome\",\n    \"zip_code\": \"00187\",\n    \"country_code\": \"RM\",\n    \"tax_domicile\": \"00111\"\n  },\n  \"fk_int_quadrature\": 0,\n  \"psp_payment\": true,\n  \"reporting_ftp\": false,\n  \"reporting_zip\": true\n}",
```
Comment (pasqualespica) [inline api-test/ApiConfig.postman_collection.json:406]: ditto

File: openapi/convert_openapi_to_swagger.sh [hunk 4]
```diff
@@ -0,0 +1 @@
+api-spec-converter  --from=openapi_3 --to=swagger_2 ./openapi.json > swagger.json
```
Comment (pasqualespica) [inline openapi/convert_openapi_to_swagger.sh:1]: add comment into script , that specify why we use this conversion and link to install locally https://www.npmjs.com/package/api-spec-converter , useful to other developers

File: pom.xml [hunk 5]
```diff
@@ -79,9 +78,8 @@
         </dependency>
 
         <dependency>
-            <groupId>javax.validation</groupId>
-            <artifactId>validation-api</artifactId>
-            <version>${validation-version}</version>
+            <groupId>org.springframework.boot</groupId>
```
Comment (pasqualespica) [inline pom.xml:81]: why this changes ?

File: src/main/java/it/pagopa/pagopa/apiconfig/ApiConfig.java [hunk 6]
```diff
@@ -5,12 +5,15 @@
...
 
+import java.util.Locale;
+
 @SpringBootApplication
 @EnableJpaRepositories
 @EnableRetry
 public class ApiConfig {
 
     public static void main(String[] args) {
+        Locale.setDefault(Locale.ENGLISH);
```
Comment (pasqualespica) [inline src/main/java/it/pagopa/pagopa/apiconfig/ApiConfig.java:16]: why added it ?

File: src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java [hunk 7]
```diff
@@ -78,6 +85,45 @@
...
+            @ApiResponse(responseCode = "404", description = "Not Found", content = @Content(mediaType = "application/json", schema = @Schema())),
+            @ApiResponse(responseCode = "409", description = "Conflict", content = @Content(mediaType = "application/json", schema = @Schema(implementation = ProblemJson.class))),
+            @ApiResponse(responseCode = "429", description = "Too many requests", content = @Content(mediaType = "application/json", schema = @Schema())),
+            @ApiResponse(responseCode = "500", description = "Service unavailable.", content = @Content(mediaType = "application/json", schema = @Schema(implementation = ProblemJson.class)))})
+    @PostMapping(value = "", produces = {"application/json"})
+    public ResponseEntity<CreditorInstitutionDetails> createCreditorInstitution(@RequestBody @Valid @NotNull CreditorInstitutionDetails creditorInstitutionDetails) {
+        return ResponseEntity.status(HttpStatus.CREATED).body(creditorInstitutionsService.createCreditorInstitution(creditorInstitutionDetails));
```
Comment (pasqualespica) [inline src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java:98]: Dis-aligned `409` ?  vs https://pagopa.atlassian.net/wiki/spaces/PAG/pages/374375153/Design+Review+API+configurazione+pagoPA _Errori
400 Bad request
404 creditorinstitutionscode not found
403 Forbidden
429 Too many requests
500 Internal Server Error_

File: src/main/java/it/pagopa/pagopa/apiconfig/controller/IcaController.java [hunk 8]
```diff
@@ -59,7 +59,7 @@
             @ApiResponse(responseCode = "500", description = "Service unavailable.", content = @Content(mediaType = "application/json", schema = @Schema(implementation = ProblemJson.class)))})
     @GetMapping(
             value = "/{idica}",
-            produces = {"application/xml"}
+            produces = {"application/xml", "application/json"}
```
Comment (pasqualespica) [inline src/main/java/it/pagopa/pagopa/apiconfig/controller/IcaController.java:62]: why ?

File: src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java [hunk 9]
```diff
@@ -78,6 +85,45 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "201", description = "OK.", content = @Content(mediaType = "application/json", schema = @Schema(implementation = CreditorInstitutionDetails.class))),
+            @ApiResponse(responseCode = "403", description = "Forbidden client error status.", content = @Content(mediaType = "application/json", schema = @Schema())),
+            @ApiResponse(responseCode = "404", description = "Not Found", content = @Content(mediaType = "application/json", schema = @Schema())),
+            @ApiResponse(responseCode = "409", description = "Conflict", content = @Content(mediaType = "application/json", schema = @Schema(implementation = ProblemJson.class))),
+            @ApiResponse(responseCode = "429", description = "Too many requests", content = @Content(mediaType = "application/json", schema = @Schema())),
+            @ApiResponse(responseCode = "500", description = "Service unavailable.", content = @Content(mediaType = "application/json", schema = @Schema(implementation = ProblemJson.class)))})
+    @PostMapping(value = "", produces = {"application/json"})
```
Comment (pasqualespica) [inline src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java:96]: if you POST `  "creditor_institution_code": ""` obtain

```
{
    "title": "INTERNAL SERVER ERROR",
    "status": 500,
    "detail": "could not execute statement; SQL [n/a]; constraint [null]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement"
}
```

File: src/main/resources/application-azure-d.properties [hunk 10]
```diff
@@ -11,11 +11,14 @@ spring.datasource.username=sa
 spring.datasource.password=sa
 spring.datasource.driver-class-name=org.h2.Driver
 
+spring.sql.init.mode=always
+spring.sql.init.data-locations=classpath:/h2/schema-h2.sql,classpath:/h2/data-h2.sql
```
Comment (pasqualespica) [inline src/main/resources/application-azure-d.properties:15]: _temporary_ we should connect to DB dev SIA too

File: src/test/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsControllerTest.java [hunk 11]
```diff
@@ -53,6 +63,34 @@ void setUp() {
...
+                .andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON));
+
+    }
+
+    @Test
+    void deleteCreditorInstitution() throws Exception {
+        mvc.perform(delete("/creditorinstitutions/1234").contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
+    }
```
Comment (pasqualespica) [inline src/test/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsControllerTest.java:93]: add some nok cases too, tnx

File: src/test/java/it/pagopa/pagopa/apiconfig/service/CreditorInstitutionsServiceTest.java [hunk 12]
```diff
@@ -119,6 +121,36 @@ void getEC_notFound() {
...
+        String expected = TestUtil.readJsonFromFile("response/update_creditorinstitution_ok.json");
+        JSONAssert.assertEquals(expected, actual, JSONCompareMode.STRICT);
+    }
+
+    @Test
+    void deleteCreditorInstitution() {
+        when(paRepository.findByIdDominio("1234")).thenReturn(Optional.of(getMockPa()));
+
+        creditorInstitutionsService.deleteCreditorInstitution("1234");
+        assertTrue(true);
```
Comment (pasqualespica) [inline src/test/java/it/pagopa/pagopa/apiconfig/service/CreditorInstitutionsServiceTest.java:151]: perhaps better so it is clearer what you want to test
```
   try {
      myClass.getBalanceForPerson(person1);
   } catch(Exception e) {
      fail("Should not have thrown any exception");
   }
```

File: src/main/java/it/pagopa/pagopa/apiconfig/controller/IcaController.java (hunk 8 above)
Comment (jacopocarlini) [reply src/main/java/it/pagopa/pagopa/apiconfig/controller/IcaController.java:62]: The file is xml, but if an error occurred ProblemJson is returned as JSON.

File: api-test/ApiConfig.postman_collection.json (hunk 1 above)
Comment (jacopocarlini) [reply api-test/ApiConfig.postman_collection.json]: I have some issues with k6. I am still trying to solve

File: src/main/java/it/pagopa/pagopa/apiconfig/ApiConfig.java (hunk 6 above)
Comment (jacopocarlini) [reply src/main/java/it/pagopa/pagopa/apiconfig/ApiConfig.java:16]: in the autogenerated response of bad request, the message was in Italian

File: pom.xml (hunk 5 above)
Comment (jacopocarlini) [reply pom.xml:81]: `spring-boot-starter-validation` **automatically** checks the input when an API is called.

File: src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java (hunk 7 above)
Comment (jacopocarlini) [reply src/main/java/it/pagopa/pagopa/apiconfig/controller/CreditorInstitutionsController.java:98]: I updated the _design review_ page

File: api-test/ApiConfig.postman_collection.json [hunk 13]
```diff
@@ -128,7 +128,13 @@
 					],
 					"request": {
 						"method": "POST",
-						"header": [],
+						"header": [
```
Comment (pasqualespica) [inline api-test/ApiConfig.postman_collection.json:131]: great
